omigami/feature_selector.py

```diff
@@ -22,68 +37,198 @@ def __init__(
-        self._results = None
+        self.results = None
+        self._minimum_features = 1
         self.post_processor = PostProcessor(robust_minimum)
-        self.executor = executor
 
-    def fit(self, X: NumpyArray, y: NumpyArray, groups: NumpyArray = None):
-        if groups is None:
-            logging.info("groups is not specified: i.i.d. samples assumed")
-            groups = np.arange(X.shape[0])
-        input_data = InputData(X=X, y=y, groups=groups)
-        outer_loop = self._make_outer_loop(input_data)
-        self._results = self._execute_repetitions(outer_loop)
-        self.selected_features = self.post_processor.select_features(self._results)
+    def fit(
+        self,
+        X: NumpyArray,
+        y: NumpyArray,
+        groups: NumpyArray = None,
+        executor: Executor = None,
+    ) -> FeatureSelector:
+        size, n_features = X.shape
+        groups = self.get_groups(groups, size)
+        input_data = InputDataset(X=X, y=y, groups=groups)
+        self.feature_evaluator.set_n_initial_features(n_features)
+
+        repetition_results = []
+
+        for _ in range(self.n_repetitions):
+            data_splitter = DataSplitter(
+                self.n_outer, self.n_inner, input_data, self.random_state,
+            )
+
+            outer_loop_results = []
+            for outer_split in data_splitter.iter_outer_splits():
+                outer_loop_result = self._deferred_run_outer_loop(
+                    input_data,
+                    outer_split,
+                    executor=executor,
+                    data_splitter=data_splitter,
+                )
+                outer_loop_results.append(outer_loop_result)
+            repetition_results.append(outer_loop_results)
+
+        self._selected_features = self._select_best_features(repetition_results)
         self.is_fit = True
         return self
 
-    def _execute_repetitions(self, outer_loop: OuterLoop) -> List[Repetition]:
-        results = []
-        for _ in range(self.repetitions):
-            outer_loop.refresh_splits()
-            result = outer_loop.run(executor=self.executor)
-            results.append(result)
-        return results
+    @staticmethod
+    def get_groups(groups: NumpyArray, size: int) -> NumpyArray:
+        if groups is None:
+            logging.info("groups is not specified: i.i.d. samples assumed")
+            groups = np.arange(size)
+        return groups
 
-    def _make_feature_evaluator(self, input_data: InputData) -> FeatureEvaluator:
-        return FeatureEvaluator(
-            input_data,
-            self.n_outer,
-            self.n_inner,
-            self.estimator,
-            self.metric,
-            self.random_state,
+    def _deferred_run_outer_loop(
+        self,
+        input_data: InputDataset,
+        outer_split: Split,
+        data_splitter: DataSplitter,
+        executor: Executor,
+    ) -> Union[Future, OuterLoopResults]:
+        if executor is None:
```

Comment:
Remove this if else here and use a custom executor similar to this one:
https://github.com/rohanpm/more-executors/blob/fc762f6720908d057177bbbd62439130e0994300/more_executors/_impl/sync.py

(we probably don't need this package just the functionality of a SyncExecutor)